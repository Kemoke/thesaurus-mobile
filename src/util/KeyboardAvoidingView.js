import React from 'react';
import {Dimensions, Keyboard, Platform, View, ViewProps, NativeModules, StatusBarIOS} from "react-native";
const {StatusBarManager} = NativeModules;

const {height: windowHeight} = Dimensions.get('window');

type KeyboardAvoidingViewProps = {
    verticalOffset?: number
} & ViewProps

export default class KeyboardAvoidingView extends React.Component<KeyboardAvoidingViewProps>{
    static defaultProps = {
        verticalOffset: 0
    };
    constructor(props) {
        super(props);
        this.state = {
            keyboardOffset: props.verticalOffset,
            statusBarHeight: 0
        };
        Keyboard.addListener('keyboardDidShow', this.handleKeyboardShow);
        Keyboard.addListener('keyboardWillShow', this.handleKeyboardShow);
        Keyboard.addListener('keyboardWillHide', this.handleKeyboardHide);
        Keyboard.addListener('keyboardDidHide', this.handleKeyboardHide);
    }

    componentWillUnmount(){
        Keyboard.removeListener('keyboardDidShow', this.handleKeyboardShow);
        Keyboard.removeListener('keyboardWillShow', this.handleKeyboardShow);
        Keyboard.removeListener('keyboardWillHide', this.handleKeyboardHide);
        Keyboard.removeListener('keyboardDidHide', this.handleKeyboardHide);
        if (Platform.OS === 'ios') this.statusBarListener.remove();
    }

    componentDidMount() {
        if (Platform.OS === 'ios') {
            StatusBarManager.getHeight((statusBarFrameData) => {
                this.setState({statusBarHeight: statusBarFrameData.height});
            });
            this.statusBarListener = StatusBarIOS.addListener('statusBarFrameWillChange', (statusBarData) => {
                this.setState({statusBarHeight: statusBarData.frame.height});
            });
        }
    }

    handleKeyboardShow = ({endCoordinates: { height }}) => {
        this.setState({keyboardOffset: height});
    };

    handleKeyboardHide = () => {
        this.setState({keyboardOffset: this.props.verticalOffset})
    };

    getKeyboardOffset = () => {
        const {keyboardOffset, statusBarHeight} = this.state;
        return Platform.select({
            ios: () => ({paddingBottom: keyboardOffset - statusBarHeight * 2}),
            android: () => ({height: windowHeight - keyboardOffset})
        })();
    };

    render() {
        const {style, ...props} = this.props;
        let viewStyle = [{height: '100%'}, this.getKeyboardOffset()];
        if (Array.isArray(style)) {
            viewStyle = [...viewStyle, ...style];
        } else {
            viewStyle.splice(0,0,style);
        }
        return (
            <View style={viewStyle} {...props}/>
        );
    }
}
