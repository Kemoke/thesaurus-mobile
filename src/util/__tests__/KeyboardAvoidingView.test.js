import React from "react";
import {shallow} from 'enzyme'
import {Dimensions, Keyboard, NativeModules, Platform, StatusBarIOS} from "react-native";
import KeyboardAvoidingView from "../KeyboardAvoidingView";
const {StatusBarManager} = NativeModules;

const {height: windowHeight} = Dimensions.get('window');

StatusBarManager.getHeight = () => ({
    height: 0
});
StatusBarIOS.addListener = () => ({remove: () => {}});

describe('KeyboardAvoidingView', function () {
    it('should render properly', function () {
        const component = shallow(
            <KeyboardAvoidingView/>
        );
        expect(component).toMatchSnapshot();
    });
    it('should render style array', function () {
        const component = shallow(
            <KeyboardAvoidingView style={[{height: 50}]}/>
        );
        expect(component).toMatchSnapshot();
    });
    it('should handle keyboard opening', function () {
        const openKeyboardListeners = [];
        const closeKeyboardListeners = [];
        Keyboard.addListener = jest.fn((name, callback) => {
            if (name === 'keyboardDidShow' || name === 'keyboardWillShow') {
                openKeyboardListeners.push(callback);
            }
            if (name === 'keyboardWillHide' || name === 'keyboardDidHide') {
                closeKeyboardListeners.push(callback);
            }
        });
        let platformTested = 'ios';
        Platform.select = config => () => config[platformTested]();
        const component = shallow(
            <KeyboardAvoidingView/>
        );
        openKeyboardListeners.forEach(callback => {
            callback({
                endCoordinates: {
                    height: 150
                }
            })
        });
        expect(component.instance().getKeyboardOffset()).toEqual({
            paddingBottom: 150
        });
        closeKeyboardListeners.forEach(callback => callback());
        expect(component.instance().getKeyboardOffset()).toEqual({
            paddingBottom: 0
        });
        platformTested = 'android';
        openKeyboardListeners.forEach(callback => {
            callback({
                endCoordinates: {
                    height: 150
                }
            })
        });
        expect(component.instance().getKeyboardOffset()).toEqual({
            height: windowHeight - 150
        });
        closeKeyboardListeners.forEach(callback => callback());
        expect(component.instance().getKeyboardOffset()).toEqual({
            height: windowHeight
        });
    });
    it('should unregister listeners on unmount', function () {
        Keyboard.removeListener = jest.fn();
        const component = shallow(
            <KeyboardAvoidingView/>
        );
        component.instance().componentWillUnmount();
        expect(Keyboard.removeListener).toHaveBeenCalledWith("keyboardDidShow", component.instance().handleKeyboardShow);
        expect(Keyboard.removeListener).toHaveBeenCalledWith("keyboardWillShow", component.instance().handleKeyboardShow);
        expect(Keyboard.removeListener).toHaveBeenCalledWith("keyboardWillHide", component.instance().handleKeyboardHide);
        expect(Keyboard.removeListener).toHaveBeenCalledWith("keyboardDidHide", component.instance().handleKeyboardHide);
    });
});