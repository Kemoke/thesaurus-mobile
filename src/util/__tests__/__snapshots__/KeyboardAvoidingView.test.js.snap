// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardAvoidingView should render properly 1`] = `
<Component
  style={
    Array [
      undefined,
      Object {
        "height": "100%",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
  verticalOffset={0}
/>
`;

exports[`KeyboardAvoidingView should render style array 1`] = `
<Component
  style={
    Array [
      Object {
        "height": "100%",
      },
      Object {
        "paddingBottom": 0,
      },
      Object {
        "height": 50,
      },
    ]
  }
  verticalOffset={0}
/>
`;
